### ISSUE TITLE
hapi: initial thoughts

### ISSUE THREAD

**Issue by tkellen** [endpoints/endpoints#14]
I really want to like it, but there are so many weird things. Here are some of them:

### No registering routes without first setting up a server connection?
[Why? I'm not sure](https://twitter.com/tkellen/status/564779878495830016). The structure I'd like to use (shown below) appears impossible:
```js
// src/index.js
const Hapi = require('hapi');
const app = new Hapi.Server();
app.route({
  type: "GET",
  handler: function (request, reply) { ... }
});
module.exports = app;

// index.js
const port = process.env.PORT || '8080';
const host = process.env.HOST || '0.0.0.0';
const server = require('./src/index');
server.connection({
  port: port,
  host: host
}).listen();
```

This is a contrived example, but I don't think a request handler should care about the port/host it is listening on. *If you put the server.connection() call before you register the route handlers, it works. Otherwise every route is a 404.*

### No affordances for mounting applications?
With express, you can compose any number of request handlers into a single server with a bunch of `server.use('/', require('path/to/thing'))` calls. Pretty direct. I'm not sure how to do that with Hapi. It must be possible, though. This is sort of an extension of the first issue... if every sub application has a host/port, how do you compose them together under a single parent?

### Plugins as mostly-handicapped middlewares?
Maybe I am suffering from "When all you have is a hammer, every problem looks like a nail.", but I want robust middleware-like functionality. The closest thing I can find for this is Hapi's [plugin system](http://hapijs.com/tutorials/plugins), which seems needlessly convoluted. Like, I don't think I even need to provide a justification for why I think the plugin system is bad, the documentation does it for me :(.

### Recommended shared global namespace for function sharing
Hapi advocates [registering shared functionality on your server instance](http://hapijs.com/tutorials/server-methods) instead of requiring the functionality you need where you need it. Everywhere I have seen this pattern, it makes for a terrible testing story. Also, this is what it winds up looking like:
https://github.com/janjarfalk/hapi-ampersand-react/blob/master/app/index.js#L22-L27
https://github.com/janjarfalk/hapi-ampersand-react/blob/master/app/methods/buildPageMarkup.js
https://github.com/janjarfalk/hapi-ampersand-react/blob/master/app/routes/route-page.js#L6-L8

Also, note the usage in route-page... doesn't this seems like something that should be a middleware?!

Sure, we could just not use this, but the fact that it is there at all gives me pause.

---

TL;DR, the only thing I like about hapi right now is the built-in validation, and we could get that very easily by using [express-joi-validator](https://www.npmjs.com/package/express-joi-validator).

**Comment by bmac** [comment]
@tkellen thanks for sharing this. I want to like hapi too but I've never gotten a chance to use it. 

## No affordances for mounting applications?
I think if you treat your plugins as your routes you can compose together multiple routes. 
https://github.com/rcorral/hapi-restful-api-example/blob/master/index.js#L21-L28
https://github.com/rcorral/hapi-restful-api-example/blob/master/routes/tasks.js. You can even supply a prefix when registering plugins to mount them under new routes.


## Plugins as mostly-handicapped middlewares?
I'm not sure if plugins are really meant to be used as middleware. This may be good and bad. I think express's over reliance on middleware can be confusing as because every middle ware touches `req` or `res` it can make it hard to track down where a property is being set in a large express app. On the other hand middleware is still useful for many case. I just wish it could be used more sparingly in express.


## Recommended shared global namespace for function sharing
I agree that this sounds like a terrible idea.

**Comment by tkellen** [comment]
Well, my first complaint is resolved: https://github.com/hapijs/hapi/pull/2397
